### ISSUE TITLE
Flutter doctor keeps saying that "Visual Studio not installed"

### ISSUE THREAD

**Issue by MuTsunTsai** [flutter/flutter#106601]
I'm facing the same issue as #55603, and here's the output of `flutter doctor -v`:

```
[√] Flutter (Channel stable, 3.0.3, on Microsoft Windows [版本 10.0.22000.739], locale zh-TW)
    • Flutter version 3.0.3 at C:\SDK\flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 676cefaaff (3 days ago), 2022-06-22 11:34:49 -0700
    • Engine revision ffe7b86a1e
    • Dart version 2.17.5
    • DevTools version 2.12.2

[√] Android toolchain - develop for Android devices (Android SDK version 30.0.3)
    • Android SDK at C:\Users\Donald\AppData\Local\Android\sdk
    • Platform android-32, build-tools 30.0.3
    • Java binary at: C:\Program Files\Android\Android Studio\jre\bin\java
    • Java version OpenJDK Runtime Environment (build 11.0.12+7-b1504.28-7817840)
    • All Android licenses accepted.

[√] Chrome - develop for the web
    • Chrome at C:\Program Files\Google\Chrome\Application\chrome.exe

[X] Visual Studio - develop for Windows
    X Visual Studio not installed; this is necessary for Windows development.
      Download at https://visualstudio.microsoft.com/downloads/.
      Please install the "Desktop development with C++" workload, including all of its default components

[√] Android Studio (version 2021.2)
    • Android Studio at C:\Program Files\Android\Android Studio
    • Flutter plugin can be installed from:
       https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
       https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 11.0.12+7-b1504.28-7817840)

[√] VS Code (version 1.68.1)
    • VS Code at C:\Users\Donald\AppData\Local\Programs\Microsoft VS Code
    • Flutter extension version 3.42.0

[√] Connected device (3 available)
    • Windows (desktop) • windows • windows-x64    • Microsoft Windows [版本 10.0.22000.739]
    • Chrome (web)      • chrome  • web-javascript • Google Chrome 103.0.5060.53
    • Edge (web)        • edge    • web-javascript • Microsoft Edge 103.0.1264.37

[√] HTTP Host Availability
    • All required HTTP hosts are available

! Doctor found issues in 1 category.
```

But I do have everything installed. Here's the proof:

![image](https://user-images.githubusercontent.com/39870664/175766318-50838085-5feb-40e1-ba54-d9df31e314a4.png)

And here's the result of running `.\vswhere -format json -utf8 -latest -prerelease -all` from the installer folder:

```
[
  {
    "instanceId": "de3c93c4",
    "installDate": "2022-06-16T02:22:23Z",
    "installationName": "VisualStudio/17.2.5+32616.157",
    "installationPath": "C:\\Program Files\\Microsoft Visual Studio\\2022\\Community",
    "installationVersion": "17.2.32616.157",
    "productId": "Microsoft.VisualStudio.Product.Community",
    "productPath": "C:\\Program Files\\Microsoft Visual Studio\\2022\\Community\\Common7\\IDE\\devenv.exe",
    "state": 4294967295,
    "isComplete": true,
    "isLaunchable": true,
    "isPrerelease": false,
    "isRebootRequired": false,
    "displayName": "Visual Studio Community 2022",
    "description": "功能強大的 IDE、免費供學生、開放原始碼參與者及個人使用",
    "channelId": "VisualStudio.17.Release",
    "channelUri": "https://aka.ms/vs/17/release/channel",
    "enginePath": "C:\\Program Files (x86)\\Microsoft Visual Studio\\Installer\\resources\\app\\ServiceHub\\Services\\Microsoft.VisualStudio.Setup.Service",
    "installedChannelUri": "https://aka.ms/vs/17/release/channel",
    "releaseNotes": "https://docs.microsoft.com/en-us/visualstudio/releases/2022/release-notes-v17.2#17.2.5",
    "thirdPartyNotices": "https://go.microsoft.com/fwlink/?LinkId=661288",
    "updateDate": "2022-06-24T07:12:00.8597685Z",
    "catalog": {
      "buildBranch": "d17.2",
      "buildVersion": "17.2.32616.157",
      "id": "VisualStudio/17.2.5+32616.157",
      "localBuild": "build-lab",
      "manifestName": "VisualStudio",
      "manifestType": "installer",
      "productDisplayVersion": "17.2.5",
      "productLine": "Dev17",
      "productLineVersion": "2022",
      "productMilestone": "RTW",
      "productMilestoneIsPreRelease": "False",
      "productName": "Visual Studio",
      "productPatchVersion": "5",
      "productPreReleaseMilestoneSuffix": "1.0",
      "productSemanticVersion": "17.2.5+32616.157",
      "requiredEngineVersion": "3.2.2159.37022"
    },
    "properties": {
      "campaignId": "976493843.1615860730",
      "channelManifestId": "VisualStudio.17.Release/17.2.5+32616.157",
      "nickname": "",
      "setupEngineFilePath": "C:\\Program Files (x86)\\Microsoft Visual Studio\\Installer\\setup.exe"
    }
  }
]
```

Please help!

**Comment by huycozy** [comment]
Hi @MuTsunTsai, thanks for filing the issue.
It seems a regression issue since it has been fixed on #39067.
Labeling the issue for further investigation from the team.

**Comment by MuTsunTsai** [comment]
@huycozy I've repeated all your steps, including completely uninstalling Visual Studio and reinstalling, without any success. The `flutter doctor -v` ends in the same output. I sincerely suggest that you think about how I may help you diagnosing this issue instead of trying to pretend that the issue isn't there based on your "personal experience" (we're not selling religion here, no).

**Comment by stuartmorgan** [comment]
> I sincerely suggest that you think about how I may help you diagnosing this issue

@MuTsunTsai The Flutter team members who do primary issue triage are not domain experts in every single part of the Flutter project. This issue was correctly triaged with the correct labels to route it to the engineers who could investigate further, as was noted above ("Labeling the issue for further investigation from the team."). Routing issues to the correct domain experts is a critical part of how the initial triage process helps diagnose issues.

> instead of trying to pretend that the issue isn't there based on your "personal experience"

Attempting to reproduce issues and documenting the results is not "pretending that the issue isn't there", it's gathering and documenting a potentially useful data point about the conditions under which an issue reproduces. It is a standard part of our triage process.

Please take a moment to read [our code of conduct](https://github.com/flutter/flutter/blob/master/CODE_OF_CONDUCT.md) before continuing to engage here.

**Comment by stuartmorgan** [comment]
> I'm facing the same issue as #55603

Since your `vswhere` output is non-empty even without `-products *`, this is unrelated to #55603/#39067.

Since your `description` has non-ASCII output, it's possible that the workaround in 3.0.2 for the buggy `vswhere` output encoding is not handling this specific case correctly in some way that causes it to be mis-parsed. (E.g., maybe there's an encoding issue with the last character in the string that causes the JSON to be invalid, rather than just that string.)

Could you:
1. pipe the output of your `vswhere` command directly to a file and then post that, so we can get the exact bytes that are being output, since copy/pasting from console may not give the same results, and then (*after* capturing that file)
2. try [this](https://github.com/flutter/flutter/issues/102451#issuecomment-1124651845) to see if it is indeed specific to incorrect `vswhere` encoding.

**Comment by MuTsunTsai** [comment]
@stuartmorgan Sure.
1. Here's my output using Powershell `Out-File` command: [vswhere.zip](https://github.com/flutter/flutter/files/8997378/vswhere.zip)
2. Yes, replacing `vswhere` solves the problem. Now `flutter doctor` finds my Visual Studio.

**Comment by stuartmorgan** [comment]
(Regardless of whether we can fix this output up enough to parse successfully, this demonstrates that we should stop silently dropping JSON parsing errors in the `vswhere` parsing, and instead error out with a clear message.)

**Comment by watamario15** [comment]
Hmm, year that looked strange but I literally typed
```pwsh
./vswhere -format json -utf8 -latest -prerelease -all > \path\to\new.txt
./vswhere_orig -format json -utf8 -latest -prerelease -all > \path\to\orig.txt
```
on PowerShell. Piping to `Out-File` made the same result. Maybe some Windows mess...

**Comment by watamario15** [comment]
Ok, I wrote a very simple program and got these files. These should be the actual output.
- [new.txt](https://github.com/flutter/flutter/files/9011921/new.txt)
- [orig.txt](https://github.com/flutter/flutter/files/9011922/orig.txt)

I leave a program that I used to get these files:
```c
#include <stdio.h>
#include <stdlib.h>
#define BUF 1024

int main(int argc, char *argv[]) {
    if(argc != 3) {
        fprintf(stderr, "Usage: %s <cmdline> <outfile>\n", argv[0]);
        return -1;
    }
    char *cmdline = argv[1], *outfile = argv[2];

    FILE *proc = popen(cmdline, "rb");
    if(!proc) {
        perror("exec");
        return -1;
    }

    FILE *fp = fopen(outfile, "wb");
    if(!fp) {
        perror("fopen");
        pclose(proc);
        return -1;
    }

    while(!feof(proc)) {
        char buf[BUF];
        size_t ret = fread(buf, sizeof(buf[0]), sizeof(buf), proc);
        fwrite(buf, sizeof(buf[0]), ret, fp);
    }

    fclose(fp);
    pclose(proc);
}
```

**Comment by stuartmorgan** [comment]
Thanks, that's perfect.

The exception with JSON parsing that we're currently dropping for that output is:
```
Unhandled exception:
FormatException: Unrecognized string escape (at line 16, character 38)
    "description": "�w���A�I�[�v���\�[�X�̋����쐬�ҁA�l�p�̖����ŋ��͂� IDE",
```

So if we want to actually handle this more comprehensively than erroring out when `vswhere` gives us code that ends up being invalid *JSON* when interpreted as UTF-8 (rather than just invalid UTF-8, which the previous PR added handling for), we'd need to do some pre-parsing fixup. E.g., we could just strip out the `description` line entirely, since it's currently the source of all the issues, and we don't actually use it.

**Comment by loic-sharma** [comment]
Hello! We're working on a fix to this issue. For now, you can unblock yourself by [switching to the "master" Flutter pre-release channel](https://docs.flutter.dev/development/tools/sdk/upgrading#switching-flutter-channels) or [updating `vswhere`](https://github.com/flutter/flutter/issues/102451#issuecomment-1124651845). Please note that the "master" channel is the bleeding edge and you may encounter bugs. We will update this thread when the fix has landed on the "stable" channel. Please let us know if you have any questions!

**Comment by loic-sharma** [comment]
Hello, the fix for this issue was released in Flutter v3.0.5. Use `flutter upgrade` to update your Flutter installation. Please let us know if you run into additional issues!

**Comment by watamario15** [comment]
I confirmed that `flutter doctor` and the Windows build is now working as expected. Thanks for your quick job!

**Comment by MuTsunTsai** [comment]
Confirm issue fixed with the original vswhere.

**Comment by Maverick-Wolf** [comment]
Hey, I update my flutter to v3.05 today and since then I have been getting the same issue.
My flutter doctor response
`[√] Flutter (Channel stable, 3.0.5, on Microsoft Windows [Version 10.0.22000.795], locale en-IN)
    • Flutter version 3.0.5 at C:\src\flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision f1875d570e (11 days ago), 2022-07-13 11:24:16 -0700
    • Engine revision e85ea0e79c
    • Dart version 2.17.6
    • DevTools version 2.12.2

[√] Android toolchain - develop for Android devices (Android SDK version 30.0.3)
    • Android SDK at C:\Users\sky\AppData\Local\Android\sdk
    • Platform android-31, build-tools 30.0.3
    • Java binary at: C:\Program Files\Android\Android Studio\jre\bin\java
    • Java version OpenJDK Runtime Environment (build 11.0.11+9-b60-7590822)
    • All Android licenses accepted.

[√] Chrome - develop for the web
    • Chrome at C:\Program Files (x86)\Google\Chrome\Application\chrome.exe

[X] Visual Studio - develop for Windows
    X Visual Studio not installed; this is necessary for Windows development.
      Download at https://visualstudio.microsoft.com/downloads/.
      Please install the "Desktop development with C++" workload, including all of its default components

[√] Android Studio (version 2021.1)
    • Android Studio at C:\Program Files\Android\Android Studio
    • Flutter plugin can be installed from:
       https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
       https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 11.0.11+9-b60-7590822)

[√] VS Code (version 1.69.2)
    • VS Code at C:\Users\sky\AppData\Local\Programs\Microsoft VS Code
    • Flutter extension version 3.44.0

[√] Connected device (3 available)
    • Windows (desktop) • windows • windows-x64    • Microsoft Windows [Version 10.0.22000.795]
    • Chrome (web)      • chrome  • web-javascript • Google Chrome 103.0.5060.134
    • Edge (web)        • edge    • web-javascript • Microsoft Edge 103.0.1264.62

[√] HTTP Host Availability
    • All required HTTP hosts are available

! Doctor found issues in 1 category.
`

**Comment by stuartmorgan** [comment]
@Maverick-Wolf Please file a new issue and include [your `vswhere` output](https://github.com/flutter/flutter/issues/55603#issuecomment-619298819).
